### PR TITLE
Make the singleton _empty immutable.

### LINKED ISSUE #158: `_empty` singleton is mutable
zope.interface.declarations returns a global `_empty` Declaration object in a few cases. But this object is fully mutable: its bases can be set, it can be subscribed and have dependents added. That doesn't seem safe for a global object. 

Can we either:

- Stop returning a global object from these APIs (return a new object); or 
- introduce a special subclass that's truly immutable (prohibits changing bases, adding dependents, etc).

### PR DESCRIPTION
**Pull request by jamadden**
Fixes #158

### CODE REVIEW

File: src/zope/interface/declarations.py [hunk 1]
```diff
@@ -111,6 +111,53 @@ def __add__(self, other):
...
+    def __reduce__(self):
+        return "_empty"
+
+    @property
+    def __bases__(self):
+        return ()
+
+    @__bases__.setter
+    def __bases__(self, new_bases):
+        if new_bases:
```
Comment (jensens) [inline src/zope/interface/declarations.py]: Why allowing setting at all? When is it the case to set an empty base here?

Comment (jamadden) [reply src/zope/interface/declarations.py]: The constructor of the base class sets the bases.

Comment (jensens) [reply src/zope/interface/declarations.py]: Ok, got it! Probably a short comment in code helps to reduce confusion of our future-selfs and others reading the code :) ?

Comment (jamadden) [reply src/zope/interface/declarations.py]: Sure!

**Pull request merged**